Sources/NIOFileSystem/IOStrategy.swift

```diff
@@ -0,0 +1,211 @@
+            // non-sensical.
+            throw FileSystemError(
+                code: .invalidArgument,
+                message: "Can't do a copy operation without at least 2 file descriptors '\(maxDescriptors)' is illegal",
+                cause: nil,
+                location: .here()
+            )
+        }
+        return .init(.parallel(maxDescriptors))
+    }
+}
+
+extension CopyStrategy: CustomStringConvertible {
+    public var description: String {
+        switch self.wrapped {
+        case .sequential:
+            return "sequential"
+        case let .parallel(maxDescriptors):
+            return "parallel with max \(maxDescriptors) descriptors"
+        }
+    }
+}
+
+/// How to perform file deletions. Currently only relevant to directory level deletions when using
+/// ``FileSystemProtocol/removeItem(at:strategy:recursively:)`` or other overloads that use the
+/// default behaviour.
+public struct RemovalStrategy: Hashable, Sendable {
+    // Avoid exposing to prevent breaking changes
+    internal enum Wrapped: Hashable, Sendable {
+        // platformDefault is reified into one of the concrete options below:
+
+        case sequential
+        // Constraints on this value are enforced only on creation of `RemovalStrategy`. The early
+        // error check is desirable over validating on downstream use.
+        case parallel(_ maxDescriptors: Int)
+    }
+
+    internal let wrapped: Wrapped
+    private init(_ strategy: IOStrategy.Wrapped) {
+        switch strategy {
+        case .sequential:
+            self.wrapped = Wrapped.sequential
+        case let .parallel(maxDescriptors):
+            self.wrapped = Wrapped.parallel(maxDescriptors)
+        }
+    }
+
+    // A deletion requires no file descriptors. We only consume a file descriptor while scanning the
+    // contents of a directory, so the minimum is 1.
+    private static let minRequiredDescriptors = 1
+
+    /// Operate in whatever manner is deemed a reasonable default for the platform. This will limit
+    /// the maximum file descriptors usage based on reasonable defaults.
+    ///
+    /// Current assumptions (which are subject to change):
+    /// - Only one delete operation would be performed at once
+    /// - The delete operation is not intended to be the primary activity on the device
+    public static let platformDefault: Self = Self(IOStrategy.determinePlatformDefault())
+
+    /// Traversal of directories and removal of files will be done sequentially without any
+    /// parallelization.
+    public static let sequential: Self = Self(.sequential)
+
+    /// Allow for one or more directory scans to run at the same time. Removal of files will happen
+    /// on asynchronous tasks in parallel.
+    ///
+    /// Setting `maxDescriptors` to 1, will limit the speed of directory discovery. Deletion of
+    /// files within that directory will run in parallel, but discovery of subdirectories will be
+    /// limited to one at a time.
```

Comment:
Writing these out, and having talked to George earlier, I wonder if we should also update `FileSystem.removeItemSequentially` to perform file deletion in parallel. All we are controlling with `RemovalStrategy` is the number of directory handles that are held by one process at a time. Deletion of items does not require depletion of file descriptors, and we could spin out additional tasks to remove items when users prefer to use `.sequential`.

Reply:
I've got changes for this locally that makes `removeItemSequentially` about 1.2x faster than the current implementation. Not a huge difference, but a simple change.

```
$ hyperfine --runs 10 --warmup 2 --prepare "rm -rf /tmp/swNIOTestCLI --sequential /tmp/swift-nio" ".build/release/NIOTestCLI --parallel /ttmp/swift-nio" "rm -rf /tmp/swift-nio"
Benchmark 1: .build/release/NIOTestCLI --sequential /tmp/swift-nio
  Time (mean ± σ):      2.298 s ±  0.065 s    [User: 0.504 s, System: 9.054 s]
  Range (min … max):    2.184 s …  2.412 s    10 runs

Benchmark 2: .build/release/NIOTestCLI --parallel /tmp/swift-nio
  Time (mean ± σ):     961.5 ms ±  57.0 ms    [User: 327.1 ms, System: 5753.0 ms
  Range (min … max):   909.1 ms … 1090.0 ms    10 runs

Benchmark 3: .build/release/NIOTestCLI --filemanager /tmp/swift-nio
  Time (mean ± σ):      1.340 s ±  0.095 s    [User: 0.021 s, System: 1.226 s]
  Range (min … max):    1.281 s …  1.599 s    10 runs

Benchmark 4: rm -rf /tmp/swift-nio
  Time (mean ± σ):      1.462 s ±  0.032 s    [User: 0.016 s, System: 1.382 s]
  Range (min … max):    1.413 s …  1.501 s    10 runs

Summary
  .build/release/NIOTestCLI --parallel /tmp/swift-nio ran
    1.39 ± 0.13 times faster than .build/release/NIOTestCLI --filemanager /tmp/swift-nio
    1.52 ± 0.10 times faster than rm -rf /tmp/swift-nio
    2.39 ± 0.16 times faster than .build/release/NIOTestCLI --sequential /tmp/swift-nio
$ tree -a . | grep directories
3768 directories, 12118 files
```

Reply:
Fair enough. Good point on users likely being surprised about this behavior when using `.sequential`.